### ISSUE TITLE
Tensorboard tf.summary.text mangles markdown linebreaks

### ISSUE THREAD

**Issue by hholst80** [tensorflow/tensorboard#222]
The following example shows how markdown with linebreaks, breaks.
The reason seems to be some zero-byte padding that happens in protobuf serialization to the underlying markdown library. The suggested fix works for us but we are unsure if this is the proper way to solve it? 

```
import tensorflow as tf
sess = tf.InteractiveSession()
text = """# Heading
\tPreformatted text,
\twith linebreaks."""
summary_op = tf.summary.text('text', tf.convert_to_tensor(text))
summary_writer = tf.summary.FileWriter('/tmp/tensorboard', sess.graph)
text = sess.run(summary_op)
summary_writer.add_summary(text, 0)
summary_writer.flush()
summary_writer.close()
```

```
def markdown_and_sanitize(markdown_string):
  """Takes a markdown string and converts it into sanitized html.

  It uses the table extension; while that's not a part of standard
  markdown, it is sure to be useful for TensorBoard users.

  The sanitizer uses the allowed_tags and attributes specified above. Mostly,
  we ensure that our standard use cases like tables and links are supported.

  Args:
    markdown_string: Markdown string to sanitize

  Returns:
    a string containing sanitized html for input markdown
  """
  # Convert to utf-8 whenever we have a binary input.
  if isinstance(markdown_string, six.binary_type):
    markdown_string = markdown_string.decode('utf-8')

  markdown_string = markdown_string.replace('\x00', '')
...
```

**Comment by wchargin** [comment]
Thanks for providing a repro script! When I run your script (verbatim except for the logdir) and launch TensorBoard from master, everything looks okay to me:
![Screenshot of the text dashboard on the results of your script](https://user-images.githubusercontent.com/4317806/28129820-a64e8466-66e9-11e7-9b93-81852d991c25.png)

Could you describe the problem that you're encountering? (Does TensorBoard fail to ingest the data? Is there a 500 when requesting the data? Does the frontend render the text incorrectly?)

**Comment by hholst80** [comment]
The data in `markdown_string` contains a lot of null-bytes otherwise it looks ok. Can you try the tensorboard that ships with 1.2.0?

**Comment by wchargin** [comment]
Sure: that would be TensorBoard 54 from TensorFlow r1.2. I get the same result. I don't see any errors in the console or in the TensorBoard logs.
![image](https://user-images.githubusercontent.com/4317806/28279744-54e759ae-6ad6-11e7-85ee-fecaed8860c9.png)

Even if I add lots of NUL characters to the input text—
```py
text = """# Heading
\tPreformatted text,
\twith linebreaks, and\0\0\0some null
\tbytes including line-final\0
\t\0and line-initial-after-tab
\0\tand line-initial-before-tab,
\tand some normal text again."""
```
—everything renders properly:
![image](https://user-images.githubusercontent.com/4317806/28280410-84f2be66-6ad8-11e7-815c-cafeed6f330d.png)

I'm still not sure that I understand your problem, or what you mean by "The data in `markdown_string` contains a lot of null-bytes otherwise it looks ok." Do you mean that the problem only occurs when the input has null bytes? (Doesn't look like it for me, and I confirmed that the null bytes are stored in the event logs: `00000340: 7265 616b 732c 2061 6e64 0000 0073 6f6d  reaks, and...som`.) Do you mean that TensorBoard shows lots of null bytes in the rendered output when there aren't actually null bytes? (I haven't observed this, and would be quite surprised.)

It's also worth noting that `bleach.clean` removes null bytes, and `markdown.markdown` passes them through unchanged, so I don't see what impact your proposed patch would have, either.

**Comment by hholst80** [comment]
A string like `'foo'` would be "scrambled" like `'\x00\x00f\x00\x00o\x00\x00o'`

**Comment by hholst80** [comment]
TF 1.2.1 on another machine gives me this result. 

```
'#\x00\x00\x00 \x00\x00\x00H\x00\x00\x00e\x00\x00\x00a\x00\x00\x00d\x00\x00\x00i\x00\x00\x00n\x00\x00\x00g\x00\x00\x00\n\x00\x00\x00\t\x00\x00\x00P\x00\x00\x00r\x00\x00\x00e\x00\x00\x00f\x00\x00\x00o\x00\x00\x00r\x00\x00\x00m\x00\x00\x00a\x00\x00\x00t\x00\x00\x00t\x00\x00\x00e\x00\x00\x00d\x00\x00\x00 \x00\x00\x00t\x00\x00\x00e\x00\x00\x00x\x00\x00\x00t\x00\x00\x00,\x00\x00\x00\n\x00\x00\x00\t\x00\x00\x00w\x00\x00\x00i\x00\x00\x00t\x00\x00\x00h\x00\x00\x00 \x00\x00\x00l\x00\x00\x00i\x00\x00\x00n\x00\x00\x00e\x00\x00\x00b\x00\x00\x00r\x00\x00\x00e\x00\x00\x00a\x00\x00\x00k\x00\x00\x00s\x00\x00\x00.\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```

**Comment by dandelionmane** [comment]
@hholst80 What system are you running on? Are you running on Windows? I feel like there may be some platform differences causing our inability to reproduce the bug you're seeing.

**Comment by hholst80** [comment]
Ubuntu 16.04 LTS and the second system was a derived (Elementary OS; but also based on 16.04 LTS)

**Comment by wchargin** [comment]
Could you please post a patch to our unit tests (say, [`text_plugin_test.py`](https://github.com/tensorflow/tensorboard/blob/b2dc1b416f748d8047befd189717eae85e7b2023/tensorboard/plugins/text/text_plugin_test.py)) such that the test fails on master but succeeds when your patch is applied?

Please be sure to run the patch in a clean virtualenv by executing the following commands verbatim:
```sh
virtualenv --no-site-packages ~/tensorflow-nightly-20170724
source ~/tensorflow-nightly-20170724/bin/activate
pip install 'https://ci.tensorflow.org/view/Nightly/job/nightly-matrix-cpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON2,label=cpu-slave/lastSuccessfulBuild/artifact/pip_test/whl/tensorflow-1.3.0rc0-cp27-none-linux_x86_64.whl'
```

Please also be sure that your git working copy is up to date with `origin/master` at the time of testing, except for the unit test that you introduced.

Thanks!

**Comment by hholst80** [comment]
I got a 404 error on the wheel file above. Sorry I was on a vacation. Can you post an new link?

**Comment by wchargin** [comment]
The following links should be stable:

  * Python 2: https://ci.tensorflow.org/view/Nightly/job/nightly-matrix-cpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON2,label=cpu-slave/lastSuccessfulBuild/artifact/pip_test/whl/tensorflow-1.head-cp27-none-linux_x86_64.whl
  * Python 3.4: https://ci.tensorflow.org/view/Nightly/job/nightly-matrix-cpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON3,label=cpu-slave/lastSuccessfulBuild/artifact/pip_test/whl/tensorflow-1.head-cp34-cp34m-linux_x86_64.whl

**Comment by wchargin** [comment]
You might also want to take a look at #239, to see if your issue is the same as is being discussed there.

**Comment by hholst80** [comment]
Yes looks like the issue I'm seeing. I am using Python 3.5.

**Comment by hholst80** [comment]
Stupid question - how do you run the test suite connected to the file above (`text_plugin_test.py`)?

**Comment by wchargin** [comment]
No worries—run `bazel test //tensorboard/plugins/text:text_plugin_test`. You may also want `bazel test //tensorboard/plugins/text:text_plugin_test --test_output=errors` so that you get feedback when there are failures.

If you've made changes that you think may warrant more general testing, you can run `bazel test //tensorboard/...` (or `//tensorboard/plugins/...`, etc.) to run all tests recursively.

**Comment by erichiller** [comment]
I'm encountering the same. Specifically with tables, regardless of line break type `\n` or `\r\n` ( or double space for that matter) it results in the same line-ending-less output.
```
| heading | heading | |--- |--- | | key | value | | key | value |
```

**Comment by achaiah** [comment]
I'm seeing the same issue with Tensorboard and markdown tables. They are not being rendered and instead appear as one single line of text (just like @erichiller has demonstrated)!  I don't have the luxury of creating a tensor since I'm publishing to Tensorboard without using TF.  Is this a bug in Tensorboard?

**Comment by zuoanqh** [comment]
I got the same problem here and it is driving me crazy. I can't believe something this trivial seems impossible. 

What kind of environment do you guys debug on? I'm using pycharm on WSL with chrome, the same code OP gives produces no linebreak, and this is disabling the table function as well. If you run the same code and results differ, it must be the environment, right? I'm using tf 1.3.0.

I've tried every solution on the internet and searched every line of code that has "tf.summary.text" on Github. To be honest I'm getting a bit frustrated. Maybe I'll just do the list trick.

**Comment by nfelt** [comment]
This sounds like it may be caused by the same null byte insertion issues that were happening in #647 and should have been recently fixed in #685.  Those appeared only for scalar strings when running python3 - is anyone here seeing this behavior with python 2 or with non-scalar tensor strings?

I can reproduce the line breaks disappearing with the OP's code when running TensorBoard under python 3 as of the commit immediately prior to #685.  Under the same conditions but at HEAD the line breaks are preserved correctly.

I'm pretty sure that this issue is fixed based on that reproduction, so I'm going to close this for now - sorry for the long wait on fixing this.  Please re-open if you encounter this problem after the next TensorBoard release or running with Python 2.x instead of 3.x.

**Comment by nfelt** [comment]
Duplicate of #647
